lib/util.h

```diff
@@ -37,6 +37,16 @@
 #define restrict
 #endif
 
+#define bug_assert(condition)                                                           \
+    if (!(condition)) {                                                                 \
+        fprintf(stderr,                                                                 \
+            "Bug detected in %s at line %d. Please open an issue on GitHub, "           \
+            "ideally with a reproducible example. "                                     \
+            "https://github.com/tskit-dev/msprime/issues\n",                            \
+            __FILE__, __LINE__);                                                        \
+        abort();                                                                        \
+    }
+
```

Comment:
There's a convention for multiline c macros that uses do-while loop. This gives you an inner brace scope, but also catches if a user calls the macro without a semicolon.
```
#define my_thing(x) do { \
  // ... do something with x \
} while(0)
```

Reply:
Ah, I was wondering if I needed that when I had an ``if`` in there. Thanks, I'll stick in the ``do`` loop.